parsl/dataflow/dflow.py

```diff
@@ -379,38 +397,83 @@ def handle_exec_update(self, task_record, future):
+            # processing will happen in another callback (on the final Future
+            # to complete)
+
+            # There is no retry handling here: inner apps are responsible for
+            # their own retrying, and joining state is responsible for passing
+            # on whatever the result of that retrying was (if any).
+
+            outer_task_id = task_record['id']
+            logger.debug(f"Join callback for task {outer_task_id}, inner_app_future {inner_app_future}")
+
+            if task_record['status'] != States.joining:
+                logger.debug(f"Join callback for task {outer_task_id} skipping because task is not in joining state")
+                return
+
+            joinable = task_record['joins']  # Future or collection of futures
+
+            if isinstance(joinable, list):  # TODO more generic type than list?
+                for future in joinable:
+                    if not future.done():
+                        logger.debug(f"A joinable future {future} is not done for task {outer_task_id} - skipping callback")
+                        return  # abandon this callback processing if joinables are not all done
+
+            # now we know each joinable Future is done
+            # so now look for any exceptions
+            e = None
+            if isinstance(joinable, Future):
+                if joinable.exception():
+                    e = joinable.exception()
+            elif isinstance(joinable, list):
+                for future in joinable:
+                    if future.exception():
+                        e = future.exception()
+            else:
+                raise TypeError(f"Unknown joinable type {type(joinable)}")
 
-        if inner_app_future.exception():
-            e = inner_app_future.exception()
-            logger.debug("Task {} failed due to failure of inner join future".format(outer_task_id))
-            # We keep the history separately, since the future itself could be
-            # tossed.
-            task_record['fail_history'].append(repr(e))
-            task_record['fail_count'] += 1
-            # no need to update the fail cost because join apps are never
-            # retried
+            if e:
+                logger.debug("Task {} failed due to failure of an inner join future".format(outer_task_id))
+                # We keep the history separately, since the future itself could be
+                # tossed.
+                task_record['fail_history'].append(repr(e))
+                task_record['fail_count'] += 1
+                # no need to update the fail cost because join apps are never
+                # retried
 
-            task_record['status'] = States.failed
-            self.tasks_failed_count += 1
-            task_record['time_returned'] = datetime.datetime.now()
-            with task_record['app_fu']._update_lock:
-                task_record['app_fu'].set_exception(e)
+                task_record['status'] = States.failed
```

Comment:
I'm not sure about the logic here, it looks like if there are multiple failed joinables the parent task is set multiple times?

Reply:
This was modelled on the behaviour of regular dependency failure handling which also has what I thought was a benign race condition.

However, on trying to reproduce that in regular dependency handling, I don't think that the regular dependency handling is benign - I can get invalid state exceptions. I'll open a separate issue for this, as it is present in master, not just this PR.

I'll also move this PR back to draft state as I'd like that issue resolved first.

Reply:
the bad dependency handling code problem is detailed and fixed here: issue #2227 and PR #2288
I still need to investigate the behaviour for joins in this PR after that fix.

Reply:
due to the join_lock around all of `handle_join_update`, and the early test in `handle_join_update` to end the callback without doing anything if the join_app task is not in `States.joining`, I think this updating behaviour will now only happen once: the first time the callback is called when all joining apps are completed.

On earlier invocations, the callback will exit earlier because of this test:
```
            if isinstance(joinable, list):
                for future in joinable:
                    if not future.done():
                        return
```

and on later invocations because of this test:

```
            if task_record['status'] != States.joining:
                return
```